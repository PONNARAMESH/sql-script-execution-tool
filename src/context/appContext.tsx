import { createTheme, MantineProvider } from "@mantine/core";

const theme = createTheme({
  /** Your theme override here */
});

export default function GlobalContext({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <MantineProvider
      // defaultColorScheme="dark"
      theme={theme}
    >
      {/* Your app here */}
      {children}
    </MantineProvider>
  );
}
