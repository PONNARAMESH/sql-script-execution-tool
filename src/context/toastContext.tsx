"use client";

import ToastsContainer from "../components/toastContainer";
import {
  TToastClosingType,
  TToastType,
} from "../components/toastContainer/toastComponent";
import { toastReducer } from "../components/toastContainer/toastReduer";
import { generateUniqueId } from "../lib/utils/helperFunction";
import { createContext, useReducer } from "react";

const initialState = {
  toasts: [],
};

export type TToastContext = {
  successToast?: (
    message: string,
    toastClosingType?: TToastClosingType,
  ) => void;
  warningToast?: (
    message: string,
    toastClosingType?: TToastClosingType,
  ) => void;
  infoToast?: (message: string, toastClosingType?: TToastClosingType) => void;
  errorToast?: (message: string, toastClosingType?: TToastClosingType) => void;
  removeToast?: (message: string, toastClosingType?: TToastClosingType) => void;
};

export const ToastContext = createContext<TToastContext>({});

export const ToastContextProvider = ({ children }: any) => {
  const [state, dispatch] = useReducer(toastReducer, initialState);

  const addToast = (
    type: TToastType,
    message: string,
    toastClosingType?: TToastClosingType,
  ) => {
    const id = generateUniqueId("TOAST");
    dispatch({
      type: "ADD_TOAST",
      payload: {
        id,
        message,
        type,
        toastClosingType: toastClosingType || "AUTO",
      },
    });
    return id;
  };

  const successToast = (
    message: string,
    toastClosingType?: TToastClosingType,
  ) => {
    return addToast("SUCCESS", message, toastClosingType);
  };

  const warningToast = (
    message: string,
    toastClosingType?: TToastClosingType,
  ) => {
    return addToast("WARNING", message, toastClosingType);
  };

  const infoToast = (message: string, toastClosingType?: TToastClosingType) => {
    return addToast("INFO", message, toastClosingType);
  };

  const errorToast = (
    message: string,
    toastClosingType?: TToastClosingType,
  ) => {
    return addToast("ERROR", message, toastClosingType);
  };

  const removeToast = (id: string) => {
    dispatch({ type: "DELETE_TOAST", payload: id });
    return id;
  };

  return (
    <ToastContext.Provider
      value={{
        successToast,
        warningToast,
        infoToast,
        errorToast,
        removeToast,
      }}
    >
      <ToastsContainer toasts={state?.toasts} position="bottomRight" />
      {children}
    </ToastContext.Provider>
  );
};
