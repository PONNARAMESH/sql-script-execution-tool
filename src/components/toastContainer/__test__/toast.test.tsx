import { render, fireEvent, screen } from "@testing-library/react";
import Toast from "../toastComponent";
import { ToastContext } from "@/context/toastContext";

// Mock styles
jest.mock("./styles.module.scss", () => ({
  successToastBGColor: "successToastBGColor",
  successIcon: "successIcon",
  success: "success",
  warningToastBGColor: "warningToastBGColor",
  warningIcon: "warningIcon",
  warning: "warning",
  warningToastFontColor: "warningToastFontColor",
  infoToastBGColor: "infoToastBGColor",
  infoIcon: "infoIcon",
  info: "info",
  errorToastBGColor: "errorToastBGColor",
  errorIcon: "errorIcon",
  error: "error",
  toast: "toast",
  toastMessage: "toastMessage",
  dismissBtn: "dismissBtn",
  toastProgress: "toastProgress",
  toastProgressBar: "toastProgressBar",
}));

// Mock Icon component
jest.mock("@wds/pace-ui", () => ({
  Icon: () => <div />,
}));

describe("Toast Component", () => {
  const removeToastMock = jest.fn();

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it("renders correctly", () => {
    const { container } = render(
      <ToastContext.Provider value={{ removeToast: removeToastMock }}>
        <Toast
          message="Success!"
          type="SUCCESS"
          id="toast1"
          toastClosingType="AUTO"
        />
      </ToastContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it("dismisses automatically after 4 seconds", () => {
    render(
      <ToastContext.Provider value={{ removeToast: removeToastMock }}>
        <Toast
          message="Success!"
          type="SUCCESS"
          id="toast1"
          toastClosingType="AUTO"
        />
      </ToastContext.Provider>,
    );
    expect(removeToastMock).toHaveBeenCalledTimes(1);
    jest.advanceTimersByTime(4000);
    expect(removeToastMock).toHaveBeenCalledWith("toast1");
  });

  it("pauses dismissal timer on mouse enter and resumes on mouse leave", () => {
    render(
      <ToastContext.Provider value={{ removeToast: removeToastMock }}>
        <Toast
          message="Warning!"
          type="WARNING"
          id="toast2"
          toastClosingType="MANUAL"
        />
      </ToastContext.Provider>,
    );

    fireEvent.mouseEnter(screen.getByTestId("toastMessage"));
    jest.advanceTimersByTime(2000); // Halfway through the timer
    expect(removeToastMock).toHaveBeenCalledTimes(2);

    fireEvent.mouseLeave(screen.getByTestId("toastMessage"));
    jest.advanceTimersByTime(2000); // Remaining time
    expect(removeToastMock).toHaveBeenCalledWith("toast1");
    expect(removeToastMock).toHaveBeenCalledWith("toast1");
  });

  it("dismisses manually when close button is clicked", () => {
    render(
      <ToastContext.Provider value={{ removeToast: removeToastMock }}>
        <Toast
          message="Error!"
          type="ERROR"
          id="toast3"
          toastClosingType="MANUAL"
        />
      </ToastContext.Provider>,
    );

    fireEvent.click(screen.getByTestId(`close-icon-toast3`));
    expect(removeToastMock).toHaveBeenCalledWith("toast3");
  });
});
