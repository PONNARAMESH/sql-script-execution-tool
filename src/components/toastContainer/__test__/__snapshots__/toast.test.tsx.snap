// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component renders correctly 1`] = `
<div>
  <div
    class="toast successToastBGColor"
    data-testid="toastMessage"
    role="menuitem"
    tabindex="0"
  >
    <div />
    <p
      class="toastMessage"
    >
      Success!
    </p>
    <button
      class="dismissBtn"
      data-testid="close-icon-toast1"
    >
      <div />
    </button>
    <div
      class="toastProgress"
    >
      <div
        class="toastProgressBar success"
      />
    </div>
  </div>
</div>
`;
